hibernate-rx-core/src/main/java/org/hibernate/rx/event/impl/DefaultRxRefreshEventListener.java

```diff
@@ -136,119 +136,126 @@ public void onRefresh(RefreshEvent event, Map refreshedAlready) throws Hibernate
+
+					LockOptions lockOptionsToUse = event.getLockOptions();
+
+					final LockMode requestedLockMode = lockOptionsToUse.getLockMode();
+					final LockMode postRefreshLockMode;
+
+					if ( e != null ) {
+						final LockMode currentLockMode = e.getLockMode();
+						if ( currentLockMode.greaterThan( requestedLockMode ) ) {
+							// the requested lock-mode is less restrictive than the current one
+							//		- pass along the current lock-mode (after accounting for WRITE)
+							lockOptionsToUse = LockOptions.copy( event.getLockOptions(), new LockOptions() );
+							if ( currentLockMode == LockMode.WRITE ||
+									currentLockMode == LockMode.PESSIMISTIC_WRITE ||
+									currentLockMode == LockMode.PESSIMISTIC_READ ) {
+								// our transaction should already hold the exclusive lock on
+								// the underlying row - so READ should be sufficient.
+								//
+								// in fact, this really holds true for any current lock-mode that indicates we
+								// hold an exclusive lock on the underlying row - but we *need* to handle
+								// WRITE specially because the Loader/Locker mechanism does not allow for WRITE
+								// locks
+								lockOptionsToUse.setLockMode( LockMode.READ );
+
+								// and prepare to reset the entry lock-mode to the previous lock mode after
+								// the refresh completes
+								postRefreshLockMode = currentLockMode;
+							}
+							else {
+								lockOptionsToUse.setLockMode( currentLockMode );
+								postRefreshLockMode = null;
+							}
+						}
+						else {
+							postRefreshLockMode = null;
+						}
+					}
+					else {
+						postRefreshLockMode = null;
+					}
+
+					return ( (RxAbstractEntityPersister) persister ).rxLoad( id, object, lockOptionsToUse, source )
+							.thenAccept(result -> {
+								if ( result!=null ) {
+
+									// apply `postRefreshLockMode`, if needed
+									if (postRefreshLockMode != null) {
+										// if we get here, there was a previous entry and we need to re-set its lock-mode
+										//		- however, the refresh operation actually creates a new entry, so get it
+										persistenceContext.getEntry(result).setLockMode(postRefreshLockMode);
+									}
+
+									// Keep the same read-only/modifiable setting for the entity that it had before refreshing;
+									// If it was transient, then set it to the default for the source.
+									if (!persister.isMutable()) {
+										// this is probably redundant; it should already be read-only
+										source.setReadOnly(result, true);
+									}
+									else {
+										source.setReadOnly(result, e == null ? source.isDefaultReadOnly() : e.isReadOnly());
+									}
+								}
+
+								UnresolvableObjectException.throwIfNull(result, id, persister.getEntityName());
+							})
+							.whenComplete( (vv,t) -> source.getLoadQueryInfluencers().setInternalFetchProfile(previousFetchProfile) );
+				} );
+	}
 
+	private CompletionStage<Void> cascadeMerge(
```

Comment:
Shouldn't this be called `cascadeRefresh`?

Reply:
Yes, sorry, thanks.

Reply:
Fixed by 68dddeeb732a81ec7c62d116b4361236b427a61a